airbyte-integrations/connectors/source-salesforce/spec.patch.json

```diff
@@ -0,0 +1,46 @@
+{
+    "required": ["credentials"],
+    "properties": {
+        "credentials": {
+            "required": [
+                "client_id",
+                "client_secret",
+                "refresh_token"
+            ],
+            "type": "object",
+            "x-oauth2-provider": "salesforce",
+            "properties": {
+                "auth_type": {
+                    "type": "string",
+                    "const": "Client",
+                    "default": "Client"
+                },
+                "client_id": {
+                    "title": "Client ID",
+                    "description": "Enter your Salesforce developer application's <a href=\"https://developer.salesforce.com/forums/?id=9062I000000DLgbQAG\">Client ID</a>",
+                    "type": "string",
+                    "order": 2,
+                    "airbyte_secret": true
+                },
+                "client_secret": {
+                    "title": "Client Secret",
+                    "description": "Enter your Salesforce developer application's <a href=\"https://developer.salesforce.com/forums/?id=9062I000000DLgbQAG\">Client secret</a>",
+                    "type": "string",
+                    "airbyte_secret": true,
+                    "order": 3
+                },
+                "refresh_token": {
+                    "title": "Refresh Token",
+                    "description": "Enter your application's <a href=\"https://developer.salesforce.com/docs/atlas.en-us.mobile_sdk.meta/mobile_sdk/oauth_refresh_token_flow.htm\">Salesforce Refresh Token</a> used for Airbyte to access your Salesforce account.",
+                    "type": "string",
+                    "airbyte_secret": true,
+                    "order": 4
+                }
+            }
+        },
+        "refresh_token": null,
```

Comment:
What does it mean to have these same top-level properties as `null` vs. the properties with the same names under `credentials`? I'm guessing it is needed for the spec mapping to work?

Reply:
@williamhbaker so if you do a patch where you set a property as `null`, you are basically asking for that field to be removed.

In this case, the connector has `refresh_token`, `client_id` and `client_secret` flat in its config, but for OAuth we require that they be under the key `credentials`. So here I'm moving those properties to `credentials` and removing the flat properties.

Then we use spec mapping to map the config keys under `credentials` back to their original flat address.